pkg/webhook/plugins/nodeaffinitywithcache/node_affinity_with_cache.go

```diff
@@ -0,0 +1,199 @@
+
+	err = injectRequiredSchedulingTerms(pod, requireRuntimes)
+	if err != nil {
+		return shouldStop, err
+	}
+
+	err = injectPreferredSchedulingTerm(pod, preferredRuntimes)
+	if err != nil {
+		return shouldStop, err
+	}
+
+	return
+}
+
+func getRequiredAndPreferredSchedulingRuntimes(pod *corev1.Pod, runtimeInfos map[string]base.RuntimeInfoInterface) (
+	requiredRuntimes map[string]base.RuntimeInfoInterface, preferredRuntimes map[string]base.RuntimeInfoInterface) {
+
+	requiredRuntimes = nil
+	preferredRuntimes = nil
+
+	// get the pod specified bound dataset
+	boundDataSetNames := map[string]bool{}
+	for key, value := range pod.Labels {
+		// if it matches, return array with length 2, and the second element is the match
+		matchString := common.LabelAnnotationPodSchedRegex.FindStringSubmatch(key)
+		if len(matchString) == 2 && value == "required" {
+			boundDataSetNames[matchString[1]] = true
+		}
+	}
+
+	// get the preferred and required runtime
+	preferredRuntimes = map[string]base.RuntimeInfoInterface{}
+	requiredRuntimes = map[string]base.RuntimeInfoInterface{}
+
+	// copy instead of use origin runtimeInfos
+	for name, runtimeInfo := range runtimeInfos {
+		preferredRuntimes[name] = runtimeInfo
+	}
+
+	for name := range boundDataSetNames {
+		// labeled dataset has no runtime(maybe wrong name), logging info instead of return error
+		runtimeInfo, ok := runtimeInfos[name]
+		if !ok {
+			log.V(1).Info("labeled dataset has no runtime")
+			continue
+		}
+
+		requiredRuntimes[name] = runtimeInfo
+		delete(preferredRuntimes, name)
+	}
+
+	return
+}
+
+func injectRequiredSchedulingTerms(pod *corev1.Pod, runtimeInfos map[string]base.RuntimeInfoInterface) error {
+	requiredSchedulingTerms := []corev1.NodeSelectorTerm{}
+	for _, runtimeInfo := range runtimeInfos {
+		requiredSchedulingTerm, err := getRequiredSchedulingTerms(runtimeInfo)
+		if err != nil {
+			return err
+		}
+		if requiredSchedulingTerms != nil {
+			requiredSchedulingTerms = append(requiredSchedulingTerms, *requiredSchedulingTerm)
+		}
+	}
+	utils.InjectNodeSelectorTerms(requiredSchedulingTerms, pod)
+
+	return nil
+}
+
+func injectPreferredSchedulingTerm(pod *corev1.Pod, runtimeInfos map[string]base.RuntimeInfoInterface) error {
+	var preferredSchedulingTerms []corev1.PreferredSchedulingTerm
+	for _, runtimeInfo := range runtimeInfos {
+		preferredSchedulingTerm, err := getPreferredSchedulingTerm(runtimeInfo)
+		if err != nil {
+			return err
+		}
+		if preferredSchedulingTerm != nil {
+			preferredSchedulingTerms = append(preferredSchedulingTerms, *preferredSchedulingTerm)
+		}
+	}
+	utils.InjectPreferredSchedulingTerms(preferredSchedulingTerms, pod)
+
+	return nil
+}
+
+func getRequiredSchedulingTerms(runtimeInfo base.RuntimeInfoInterface) (requiredSchedulingTerm *corev1.NodeSelectorTerm, err error) {
+	requiredSchedulingTerm = nil
+	if runtimeInfo == nil {
+		err = fmt.Errorf("RuntimeInfo is nil")
+		return
+	}
+
+	requiredSchedulingTerm = &corev1.NodeSelectorTerm{
```

Comment:
Maybe it's better to check if the `runtimeInfo` is in fuse global mode. In fuse non-global mode, pod with dataset pvc will be put aside with the cache worker pod according to pvc nodeSelector. So, I think no need to inject requiredSelectorTerm.

Reply:
I think it's fine to inject requiredSelectorTerm because it can make the logic easy to understand, and I think `no-global` mode will be deprecated because it's a concept which is complex to the end user.